### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by jsilvestre** [cozy/cozy-emails#121]
We talked about the laggy feel we have when we use the application. Here are some ideas to fix it the right way.

1. Measure: what takes times, what doesn't
2. Prioritize: Pareto has an awesome rule for that (80/20)
3. Measure
4. Repeat

I suggest to tackle those areas:
* how much time does React spend in rendering for nothing (data has not changed)?
* how much time does React spend actually rendering stuff (data has changed)?
* how much time does it take to execute the `shouldComponentUpdate` (underscore's deep equal is slow)?
* are the stores' getters (read) fast?
* are the store' handlers (write) fast?
* how much time the application is render for a route change (e.g. open an email)?
* are there routes slower than others?

**Comment by jsilvestre** [comment]
## How to measure?
* http://facebook.github.io/react/docs/perf.html
* we could give https://github.com/paulirish/automated-chrome-profiling#readme a try
* I don't know if something like:
```coffeescript
start = Date.now()
...
...
..
console.log Date.now() - start
```
Can be relevant.

**Comment by jsilvestre** [comment]
# Use case: opening an email
## What is measured
* go to / (default mailbox)
* open an email

## Results

### All lifecycle time
![read_emails_inclusive](https://cloud.githubusercontent.com/assets/568111/6124878/6ff105b4-b111-11e4-9ac8-509bac4e9ed6.png)

### Wasted time
![read_emails_wasted](https://cloud.githubusercontent.com/assets/568111/6124880/75bbebee-b111-11e4-969b-fe2168016c7a.png)

Measures are relatively consistent in scale and priority order (when I repeat the whole measure process multiple times).
This result seems normal because there is no `shouldComponentUpdate` method for components `MessagesItem`, `Participants`.

**Comment by jsilvestre** [comment]
I did the improvement in my previous comment, and noticed the new bottleneck was the Conversation/Message components. They both have shouldComponentUpdate, so I digged further.

It appears that disabling HTML view in the options improves the performance a lot.
See with HTML disabled:
![no_html](https://cloud.githubusercontent.com/assets/568111/6215514/9e111bd8-b601-11e4-9190-7e7e7f0abc20.png)

With HTML enabled:
![with_html](https://cloud.githubusercontent.com/assets/568111/6215515/9e15626a-b601-11e4-9062-906c44bb8bfc.png)

The factor is almost 1,9. I noticed the HTML generation was done client-side every time the message is displayed. It means that:
* parsing text to markdown if it's possible
* responses embedding (regex to put previous emails of the conversation in a separate div with border-left in blue or red)
Is done every single time. That's probably one of the causes of slow message displaying on mobile, maybe the main one.

One possible solution would be to do what I mentioned earlier server-side, once, when the email is imported. What do you think? I'm not sure I measured that correctly, but I don't see another reason for the wasted time, `shouldComponentUpdate` always return false :-/

**Comment by clochix** [comment]
I agree that message manipulation should be done only once and server-side.
The response embedding algorithm is pretty buggy for now (see #147), so I would suggest to perform it server side but not save it so we can fix it (or save it to another field, like prettified_html)

**Comment by poupotte** [comment]
Performances are better.
However, I keep this issue opened to have a note for next improvements.
